File: Java/MaxIslandPath.java
/*Given a 2D grid of '1's (land) and '0's (water), find the max path better the islands
Only vertical and horizontal paths are valid & 1 signifies an island  & 0 signifies water
*/

class MaxIslandPath {
    int rows;
    int cols;

    public int DFS(int[][] map, int row, int col){

        // In bounds & is island check
        if(row < 0 || row >= rows || col < 0 || col >= cols || map[row][col] != 1) {return 0;}

        // Mark current as visited & check neighbors
        map[row][col] = -1;
        int size = 1;
        int[] dirRows = {1, -1, 0, 0};
        int[] dirCols = {0, 0, 1, -1};
        for(int i = 0; i < 4; i++){
            size += DFS(map, row + dirRows[i], col + dirCols[i]);
        }

        return size;
    }

    public int maxIslandPath(int[][] map){
        if(map == null || map.length == 0) {return 0;}
        rows = map.length;
        cols = map[0].length;
        int max = 0;

        for(int row = 0; row < rows; row++){
            for(int col = 0; col < cols; col++){
                if(map[row][col] == 1){
                    int size = DFS(map, row, col);
                    max = Math.max(max, size);
                }
            }
        }

        return max;
    }

    public static void main(String[] args){
        MaxIslandPath maxPath = new MaxIslandPath();
        int[][] map = {
            {0, 1, 0, 0},
            {0, 1, 1, 0},
            {0, 0, 0, 1}
        };
        System.out.println("Max Path: " + maxPath.maxIslandPath(map));
    }
}

File: Python/MaxIslandPath.py
# Given a 2D grid of '1's (land) and '0's (water), find the max path better the islands
# Only vertical and horizontal paths are valid & 1 signifies an island  & 0 signifies water
# Depth First Search Strategy

class MaxIslandPathProblem:
    def __init__(self):
        self.rows = 0
        self.cols = 0

    def DFS(self, grid, row, col):
        # In bounds & is island check
        if (
            row < 0 or row >= self.rows or
            col < 0 or col >= self.cols or
            grid[row][col] != 1
        ):
            return 0

        # Mark current as visited & check neighbors
        grid[row][col] = -1

        size = 1
        directions = [(1,0), (-1,0), (0,1), (0,-1)]

        for dr, dc in directions:
            size += self.DFS(grid, row + dr, col + dc)

        return size

    def MaxIslandPath(self, map):
        if not map or not map[0]:
            return 0

        self.rows = len(map)
        self.cols = len(map[0])
        maxPath = 0

        for r in range(self.rows):
            for c in range(self.cols):
                if map[r][c] == 1:
                    maxPath = max(maxPath, self.DFS(map, r, c))

        return maxPath

def main():
    solver = MaxIslandPathProblem()
    map = [
        [0, 1, 0, 0],
        [0, 1, 1, 0],
        [0, 0, 0, 0],
        [0, 0, 0, 1]
    ]
    print("Max Path:", solver.MaxIslandPath(map))


if __name__ == "__main__":
    main()